builder/builder.go

```diff
@@ -290,71 +291,107 @@ func (b *Builder) parseWithdrawalMessages(
+// message nonce used for the withdrawal to the tx events. This is used for proving withdrawals.
 func (b *Builder) storeWithdrawalMsgInEVM(
-	withdrawalMsg *rolluptypes.MsgInitiateWithdrawal,
+	execTxResult *abcitypes.ExecTxResult,
 	ethState *state.StateDB,
 	header *monomer.Header,
-) (*big.Int, error) {
+) error {
 	monomerEVM, err := evm.NewEVM(ethState, header)
 	if err != nil {
-		return nil, fmt.Errorf("new EVM: %v", err)
+		return fmt.Errorf("new EVM: %v", err)
 	}
 	executer, err := bindings.NewL2ToL1MessagePasserExecuter(monomerEVM)
 	if err != nil {
-		return nil, fmt.Errorf("new L2ToL1MessagePasserExecuter: %v", err)
+		return fmt.Errorf("new L2ToL1MessagePasserExecuter: %v", err)
 	}
 
 	// Get the current message nonce before initiating the withdrawal.
 	messageNonce, err := executer.GetMessageNonce()
 	if err != nil {
-		return nil, fmt.Errorf("get message nonce: %v", err)
+		return fmt.Errorf("get message nonce: %v", err)
 	}
 
-	senderCosmosAddress, err := sdk.AccAddressFromBech32(withdrawalMsg.GetSender())
+	var event *abcitypes.Event
+	for i := range execTxResult.Events {
+		event = &execTxResult.Events[i]
+		if event.Type == rolluptypes.EventTypeWithdrawalInitiated {
+			break
+		}
+	}
+	if event == nil {
+		return fmt.Errorf("withdrawal event not found")
+	}
+
+	params, err := parseWithdrawalEventAttributes(event)
 	if err != nil {
-		return nil, fmt.Errorf("convert sender to cosmos address: %v", err)
+		return fmt.Errorf("parse withdrawal attributes: %v", err)
 	}
 
 	// Initiate the withdrawal in the Monomer ethereum state.
-	if err = executer.InitiateWithdrawal(
-		common.BytesToAddress(senderCosmosAddress.Bytes()),
-		withdrawalMsg.Value.BigInt(),
-		common.HexToAddress(withdrawalMsg.GetTarget()),
-		new(big.Int).SetBytes(withdrawalMsg.GasLimit),
-		withdrawalMsg.GetData(),
-	); err != nil {
-		return nil, fmt.Errorf("initiate withdrawal: %v", err)
+	if err = executer.InitiateWithdrawal(*params); err != nil {
+		return fmt.Errorf("initiate withdrawal: %v", err)
 	}
 
-	return messageNonce, nil
+	// Populate the nonce in the tx event attributes.
+	event.Attributes = append(event.Attributes, abcitypes.EventAttribute{
+		Key:   rolluptypes.AttributeKeyNonce,
+		Value: hexutil.Encode(messageNonce.Bytes()),
+	})
+
+	return nil
+}
+
+// parseWithdrawalEventAttributes parses the withdrawal event attributes and returns the converted withdrawal parameters.
+func parseWithdrawalEventAttributes(withdrawalEvent *abcitypes.Event) (*crossdomain.Withdrawal, error) {
+	var params crossdomain.Withdrawal
+	for _, attr := range withdrawalEvent.Attributes {
+		switch attr.Key {
+		case rolluptypes.AttributeKeySender:
+			senderCosmosAddress, err := sdk.AccAddressFromBech32(attr.Value)
+			if err != nil {
+				return nil, fmt.Errorf("convert sender to cosmos address: %v", err)
+			}
+			// TODO: use updated cosmos -> eth address conversion
+			sender := common.BytesToAddress(senderCosmosAddress.Bytes())
```

Comment:
Is this a TODO?

Reply:
Ah man yeah I forgot to remove the TODO, thanks for pointing this out. I'll remove it